arch/risc-v/src/common/generated/riscv_cpux_idlestack.in

```diff
@@ -0,0 +1,32 @@
+/****************************************************************************
+ * arch/risc-v/src/common/riscv_cpux_idlestack.in
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include "riscv_internal.h"
+#include "chip.h"
+
+const uint8_t * const g_cpux_idlestack[CONFIG_SMP_NCPUS] = {
+  @RV_CPUX_IDLE_STACK@
```

Comment:
Why not define g_cpux_idlestack directly?
```
uint8_t aligned_data(16) g_cpux_idlestack[CONFIG_SMP_NCPUS][SMP_STACK_SIZE];
```

Reply:
Define this directly will simple increase the output binary. As allocate idle stack dynamically will provide the same interface but with a smaller output.

Reply:
So, are you evaluating code size of nuttx? However, the introduction of generation strategies will reduce code readability

Reply:
This is a test result.

With this patch:
``` 
  text	   data	    bss	    dec	    hex	filename
 164211	    365	   9760	 174336	  2a900	nuttx
```

With your suggestion.
```
   text	   data	    bss	    dec	    hex	filename
 164213	    365	  11792	 176370	  2b0f2	nuttx
```

> However, the introduction of generation strategies will reduce code readability

I agree with your opinion, dynamic allocation does make code harder to understand. But I do not think it is too hard to understand. In fact, I have done something to improve the readability, `riscv_set_inital_sp` give a clear view on how to setup idle stack.  It just is `base + size * hartid`. This allocation is quite simple, right?

Reply:
why not unify g_idle_topstack as the single core case as:
```
const uintptr_t g_idle_topstack = (uintptr_t)_ebss +
    CONFIG_SMP_NCPUS * CONFIG_IDLETHREAD_STACKSIZE;
```
it's easy to compute the per cpu stack from g_idle_topstack.

Reply:
> why not unify g_idle_topstack as the single core case as:
> 
> ```
> const uintptr_t g_idle_topstack = (uintptr_t)_ebss +
>     CONFIG_SMP_NCPUS * CONFIG_IDLETHREAD_STACKSIZE;
> ```
> 

Use _ebss will break bl602 (It is too hard for me to change its memory layout). I think it will be better to use the macro `RV_IDLESTACK_BASE` this patch provides.

```
const uintptr_t g_idle_topstack = RV_IDLESTACK_BASE +
    CONFIG_SMP_NCPUS * CONFIG_IDLETHREAD_STACKSIZE;
```

> it's easy to compute the per cpu stack from g_idle_topstack.

I am not sure it is OK to add this change to this patch. As you already mentioned, I will add this.

Reply:
> I agree with your opinion, dynamic allocation does make code harder to understand. But I do not think it is too hard to understand. In fact, I have done something to improve the readability, `riscv_set_inital_sp` give a clear view on how to setup idle stack. It just is `base + size * hartid`. This allocation is quite simple, right?

Yeah, so I prefer the code you implemented before, I was just wondering if there is a cleaner implementation other than introducing generated code

Reply:
> Yeah, so I prefer the code you implemented before, I was just wondering if there is a cleaner implementation other than introducing generated code

At now, I only find it is possible to implement static idle stack pointer by generating codes, other way may still need ugly macro check :(
I even hope there is a better way so I can drop this and implement this elegantly.

Reply:
but macro check is better than the dynamic code generation.

Reply:
> > Yeah, so I prefer the code you implemented before, I was just wondering if there is a cleaner implementation other than introducing generated code
> 
> At now, I only find it is possible to implement static idle stack pointer by generating codes, other way may still need ugly macro check :( I even hope there is a better way so I can drop this and implement this elegantly.

but macro check is better than the dynamic code generation.

Reply:
> > why not unify g_idle_topstack as the single core case as:
> > ```
> > const uintptr_t g_idle_topstack = (uintptr_t)_ebss +
> >     CONFIG_SMP_NCPUS * CONFIG_IDLETHREAD_STACKSIZE;
> > ```
> 
> Use _ebss will break bl602 (It is too hard for me to change its memory layout). If anyone can reorder its memory layout, it is fine to just use _ebss. At now, I think it will be better to use the macro `RV_IDLESTACK_BASE` this patch provides.

bl602 is a single core chip, it shouldn't hard to reserve some space after data/bss section for idle stack.

Reply:
> > > why not unify g_idle_topstack as the single core case as:
> > > ```
> > > const uintptr_t g_idle_topstack = (uintptr_t)_ebss +
> > >     CONFIG_SMP_NCPUS * CONFIG_IDLETHREAD_STACKSIZE;
> > > ```
> > 
> > 
> > Use _ebss will break bl602 (It is too hard for me to change its memory layout). If anyone can reorder its memory layout, it is fine to just use _ebss. At now, I think it will be better to use the macro `RV_IDLESTACK_BASE` this patch provides.
> 
> bl602 is a single core chip, it shouldn't hard to reserve some space after data/bss section for idle stack.

Yes, it is single core. But I am not sure whether there is some side effect after moving it. Is there anyone can help and give a test? I don't have a bl602 board. 
https://github.com/apache/nuttx/blob/master/boards/risc-v/bl602/bl602evb/scripts/ld.script#L167-L171

Reply:
we have the board to test the change.

Reply:
@xiaoxiang781216 Excellent, I open a new PR #12288 for this.